Fix bot shutdown hang on connection error (or other unhandled exceptions)

CyberKiller opened the pull request:
### Description of the changes
Change red_exception_handler() from calling shutdown_handler() to sys.exit()
fixes #5780 

### Have the changes in this PR been tested?
Yes

CyberKiller committed:
Fix bot shutdown hang on connection error

Added a catch all exception to the try block where red.start() is called to ensure bot shuts down correctly.

CyberKiller committed:
Fix Indent

CyberKiller committed:
Added a comment

CyberKiller commented:
On further thought, I realise that a catch all exception will probably not be acceptable, as it will catch exceptions during the running of the bot and not just at start up so I am going to convert this pull request to a draft and look for a better solution to this.

Jackenmen commented:
Errors while the bot is running don't generally propagate this far and are handled earlier by things such as loop's exception handler. If they do propagate here, something more serious happened and we do probably want to shutdown after. If connection errors propagate here then the placement is likely fine.

CyberKiller commented:
> Errors while the bot is running don't generally propagate this far and are handled earlier by things such as loop's exception handler. If they do propagate here, something more serious happened and we do probably want to shutdown after. If connection errors propagate here then the placement is likely fine.

So do you think a catch all exception is appropriate or would catching more specific exceptions be better?

Jackenmen commented:
It may be fine, can't really comment on that until I have a chance to test/play with it.

CyberKiller commented:
I suspect if any exception doesn't get caught by run_bot() while the bot is starting up then the hang will occur. ~~The question is does having a catch all exception that calls sys.exit() prevent shutdown_handler() from doing what it needs to do?~~

CyberKiller committed:
Revert "Added a comment"

This reverts commit c95a21097e832eed60f78a4b038d0989afcb630b.

Revert "Fix Indent"

This reverts commit 7605ce3320af0f26b655172813194b10219a2b24.

Revert "Fix bot shutdown hang on connection error"

This reverts commit 434e747f5c222907a4bd322d5ae77f59e6f2994e.

CyberKiller committed:
A better fix for the hanging

red_exception_handler() hangs if it tries to call shutdown_handler() it's self so just call sys.exit() instead

CyberKiller commented:
I had a little more of a poke around and I think I found out the real cause of the hang: if red_exception_handler() catches an exception and calls shutdown_handler(), then the code never falls through back into main().
Replacing the call to shutdown_handler() in red_exception_handler() with sys.exit() allows the code to fall back though to main() and allows it to call the shutdown_handler() and run the finally block at the end of main() therefore stopping the loop correctly without hanging.
As for why it was behaving in this way I have no idea.
